Migrate existing commands to a centralized console

cedric-anne opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | see #4592

![image](https://user-images.githubusercontent.com/33253653/45629415-b0456a00-ba96-11e8-869c-597e325f9d83.png)

Here is a portage of main `scripts/*.php` commands on the `symfony/command` component.

cedric-anne commented:
Some improvements that can be done :
1. Instead of having `--dev` and/or `--force`, it may be possible to interact with user.
2. Clean the CliMigration classe to limit usage of ob_*.

I also need to remove the current cliupdate script and update the documentation if solution is approved.

trasher commented:
This should traget master, not bugfixes branch (huge change).

For the first point, I  disagree to replace dev and force switches, they should be kept. Also, I'm not sure adding interaction with the user on those ones is really a good idea; because those options should not be used for most of the cases (only dev/testers -- who should know what they do) may use those.

cedric-anne commented:
I updated the code:

1. Rename command (`glpi:database:update` -> `glpi` common prefix [will be usefull if we include commands from external libs], `database` section [will be usefull to find it easier if we make a lot of commands)
2. Proxy old command to new command (to use a deprecation state before removal)
3. Add (optionnal) interaction
4. Rename parameters (to be more relevant for non dev users)
5. Refactored and comment code
6. Style output
7. Put strings into translation

cedric-anne commented:
@trasher I added interaction only to ask confirmation after displaying current/target versions. For interaction on dev/force case, I agree with you, we should not push users into those usecases.

trasher commented on inc/console/abstractcommand.class.php:
```diff
@@ -0,0 +1,110 @@
+      $GLPI = new GLPI();
+      $GLPI->initLogger();
+
+      \Config::detectRootDoc();
```
`\` not needed

cedric-anne replied:
fixed

trasher commented on inc/console/database/updatecommand.class.php:
```diff
@@ -0,0 +1,195 @@
+         'force',
+         'f',
+         InputOption::VALUE_NONE,
+         'Force execution of update from v-1 version of GLPI even if schema does not changed'
```
"did not change"?

trasher commented on inc/console/database/updatecommand.class.php:
```diff
@@ -0,0 +1,195 @@
+      $this->addOption(
+         'lang',
+         'l',
+         InputOption::VALUE_OPTIONAL
```
Maybe a line for help?

trasher commented on inc/console/database/updatecommand.class.php:
```diff
@@ -0,0 +1,195 @@
+      $current_version     = $currents['version'];
+      $current_db_version  = $currents['dbversion'];
+
+      global $migration;
```
why a global?

cedric-anne replied:
It is used in update_XX_YY functions.

trasher replied:
oh... right :/

trasher commented on inc/console/database/updatecommand.class.php:
```diff
@@ -0,0 +1,195 @@
+      } else if ($force) {
+         // Replay last update script even if there is no schema change.
+         // It can be used in dev environment when update script has been updated/fixed.
+         include_once(GLPI_ROOT . '/install/update_93_94.php');
```
Out of the scope of the current PR, but if wa can find a way to get rid of those lines to change manually on each update :(

cedric-anne replied:
Maybe in another PR. It could also be a good idea to remove `case GLPI_PREVER:` in classic update process and move it into a `Update::replayLast()` method or something like that.

trasher commented on scripts/cliupdate.php:
```diff
@@ -30,139 +30,35 @@
 
-if (in_array('--help', $_SERVER['argv'])) {
-   die("usage: ".$_SERVER['argv'][0]."  [ --force ] [ --lang=xx_XX ] [ --config-dir=/path/relative/to/script ] [--dev]\n");
+echo('Usage of this script has been deprecated, please use ../console.php glpi:database:update.' . PHP_EOL);
```
Hum... Most of the time, cliupdate is called from glpi root directory; so `../` may be confusing

cedric-anne replied:
As we see together, the old cliupdate script has been removed.

cedric-anne commented:
@trasher 
I added a shebang and autodetection of commands in core and plugins.
If it is OK for you, I will update the changelog and squash commit.

trasher commented:
Yep, go ;)

trasher commented:
`CommandFinder` should be tested :)

cedric-anne commented:
`CommandFinder` is tested.

cedric-anne commented on inc/console/application.class.php:
```diff
@@ -36,28 +36,92 @@
+    *
+    * @var string
+    */
+   const DEFAULT_CONFIG_DIR = GLPI_ROOT . '/config';
```
As it is used only once, I removed It.

cedric-anne commented:
My last commit add ability to translate commands descriptions.

The only parts that will not be translated are the default commands `help` and `list`, and labels surrounded in following image:
![console](https://user-images.githubusercontent.com/33253653/46794152-84d93680-cd47-11e8-8568-8523aad28e24.png)

trasher commented on inc/console/application.class.php:
```diff
@@ -36,58 +36,96 @@
+               null,
+               InputOption::VALUE_OPTIONAL,
+               __('Configuration directory to use'),
+               GLPI_ROOT . '/config'
```
Should be `GLPI_CONFIG_DIR`

cedric-anne commented on inc/console/task/unlockcommand.class.php:
```diff
@@ -0,0 +1,161 @@
+      $unlocked_count = 0;
+
+      foreach ($task_iterator as $task) {
+         if (!$all && !in_array($task['name'], $tasks)) {
```
Potentially, the name could ambiguous. I think we should change this argument to use a format like `itemtype.name` that cannot be ambiguous as the unicity key is buld on these columns.

cedric-anne replied:
Done.

cedric-anne commented on inc/console/migration/buildmissingtimestampscommand.class.php:
```diff
@@ -0,0 +1,155 @@
+
+         $result = $this->db->query(
+            "
+            UPDATE `$table`
```
It is not possible to not use RAW query here.

cedric-anne commented on inc/console/migration/buildmissingtimestampscommand.class.php:
```diff
@@ -0,0 +1,155 @@
+
+      global $CFG_GLPI;
+
+      $types = array_merge(
```
I just get existing list (and removed `NetworkEquipmentFirmware` that does not exists anymore). I am not really sure it is up to date.

Maybe we should just loop on all tables (except `glpi_logs`) and update them if fields `date_creation` and/or `date_mod` are found.

cedric-anne replied:
Done in 4f9bd20b72645d3353cb7a8fe6de5bf77fdd25f4.

cedric-anne commented:
This PR can be reviewed now.

I made one commit per command + one commit for the console. Could be easilly splitted in multiple PR (the only shared code between commands is the AbstractCommand class).

Other commands will be done later in separate PR.

trasher commented:
Unit tests are failing ;)

It's OK for me to have only one PR with those basics commands.

cedric-anne commented:
Rebase done and tests fixed.
